src/storage-controller/src/collection_mgmt.rs

```diff
@@ -1153,6 +1049,55 @@ where
     (tx, handle.abort_on_drop(), shutdown_tx)
 }
 
+async fn monotonic_append<T: Timestamp + Lattice + Codec64 + TimestampManipulation>(
+    write_handle: &mut WriteHandle<SourceData, (), T, Diff>,
+    updates: Vec<TimestamplessUpdate>,
+    at_least: T,
+) {
+    let mut current_upper = write_handle.shared_upper();
+    loop {
+        if updates.is_empty() && current_upper.is_empty() {
+            // Ignore timestamp advancement for
+            // closed collections. TODO? Make this a
+            // correctable error
+            return;
+        }
+
+        let upper = current_upper
+            .into_option()
+            .expect("cannot append data to closed collection");
+
+        let lower = if upper.less_than(&at_least) {
+            at_least.clone()
+        } else {
+            upper.clone()
+        };
+
+        let upper = TimestampManipulation::step_forward(&lower);
+        let updates = updates
+            .iter()
+            .map(|TimestamplessUpdate { row, diff }| {
+                ((SourceData(Ok(row.clone())), ()), lower.clone(), diff)
+            })
+            .collect::<Vec<_>>();
+        let res = write_handle
+            .compare_and_append(
+                updates,
+                Antichain::from_elem(lower),
```

Comment:
This is where it differs from the original impl, which uses `current_upper` as the expected upper and only uses `lower` to determine a timestamp for the timestampless updates.

As is here, I would expect this to fail when `at_least` is higher than `current_upper`, because the `compare_and_append` would then fail because the expected upper doesn't line up.

Reply:
Interesting! I'll have to load this up again to feel like I totally understand that, but sounds like a pretty mechanical fix.

Something I don't understand is that previously, the `MonotonicAppend` call could return a `StorageError` that the upper didn't match. But why? Isn't the contract just that the thing gets written down at some timestamp that's at least the `at_least`? Do you know if this was just an artifact of implementation (assigning a timestamp and then afterward treating it like a normal Append) or is that necessary for some reason?

Reply:
I think the intention was "we pick a timestamp that's guaranteed to increase, but we still do a compare_and_append to ensure that there's no other writer". I also believe that the reasoning around these "monotonic appends" is a bit hand-wave-y, especially when you consider potential concurrent writers, in a use-case isolation future, and just in general. It comes from a world where we didn't have the differential collections and from a world where we didn't think about 0dt/uci and relied a lot on "yeah yeah, `envd` fencing makes this all correct" (which I never fully trusted... 😅 ).

So it isn't what I would call a _blind write_, where the semantics really are: write down the thing, no matter what, if there's a concurrent write and the upper doesn't match you can retry.

It's part of the reason I introduced the differential collections, which have a sync-and-retry loop and generally feel much more robust. And now that the "partially truncate" logic also used a proper compare_and_append (I changed that in [1]), I feel better about our consistency story. And it might be that all users of `monotonic_append` now are okay with them being blind writes.

[1]:
https://github.com/MaterializeInc/materialize/blob/f5b681abad68572d46b650a11f5196ae3d39eea4/src/storage-controller/src/lib.rs#L3343